contributors/design-proposals/bootstrap-discovery.md

```diff
@@ -173,6 +172,47 @@ In addition, `jws-kubeconfig-<token-id>-hash` will be set to the MD5 hash of the
 
 This controller will also delete tokens that are past their expiration time.
 
+## `kubeadm` UX
+
+We extend kubeadm with a set of flags and helper commands for managing and using these tokens.
+
+### `kubeadm init` flags
+
+* `--token` If set, this injects the bootstrap token to use when initializing the cluster.  If this is unset, then a random token is created and shown to the user.  If set explicitly to the empty string then no token is generated or created.  This token is used for both discovery and TLS bootstrap.
+* `--token-ttl` If set, this sets the TTL for the lifetime of this token.  Defaults to 0 which means "forever"
+
+### `kubeadm join` flags
+
+* `--token` This sets the token for both discovery and bootstrap auth.
+* `--discovery-url` If set this will grab the cluster-info data (a kubeconfig) from a URL. Due to the sensitive nature of this data, we will only support https URLs.  This also supports `username:password@host` syntax for doing HTTP auth.
+* `--discovery-file` If set, this will load the cluster-info from a file.
+* `--discovery-token` If set, (or set via `--token`) then we will be using the token scheme described above.
```

Comment:
Wrt Discovery -- This is all outlined in the first section of the document.  This is about distributing information about the cluster itself including the CA bundle for that server.  Over time we can use this to find a set of API servers but that is for the future.  This has nothing to do with auth at all and it is bad news if you put TLS client certificates in cluster-info.  Earlier in the doc I say "It is expected that the `kubeconfig` file will have a single unnamed `Cluster` entry.  Other information (especially authentication secrets) must be omitted."

But, once you find a cluster, how do you get a unique client certificate signed by a CA that the server trusts?  That is the TLS bootstrap flow.  And the option we are using here os to also use the token for that.  If you have a client certificate that you've already gotten via other means we should support that.  But that really isn't "bootstrap" in any way.

The client TLS stuff is a bit out of scope for this doc but I'd be cool with a set of flags like this:

* `--client-tls-bootstrap-token` Configure client TLS by using a token and the Cert API
* `--client-tls-certificate` and `--client-tls-key` Instead of doing TLS bootstrapping with a token, read the client certificate and key from these files.  Install them in the appropriate directories.  Verify that they can be used to talk to the server.  Note that the flag names mirror the `kubectl` flags somewhat.

If either `--client-tls-certificate` or `--client-tls-key` are set then it is assumed that users do not want to use the cert API to bootstrap TLS.  If both are not set or they cannot successfully be used to communicate with the server then an error is thrown.